### DISCUSSION TITLE
Conversion from markdown to PDF results in "Failed to load file", even with "--resource-path" flag

### DISCUSSION THREAD

**Discussion by impostorsguides** [jgm/pandoc#9315] [General]
I have a Jekyll site that I publish on Github pages, and the directory structure looks like so:

```
/main_directory/
    -/assets/
        -/images/
            -2024-01-06-1037am.png
            -screenshot-2024-01-05-10-23-46.png
    -/rbenv/
        -/part-0/
            -install-rbenv-content.md
```

I want to convert `install-rbenv-content.md` to a PDF file, *including* its embedded images.  The images are included in the markdown file like so:

```
![some alt text](/assets/images/screenshot-2024-01-05-10-23-46.png)
```

This loads fine when I spin up the Jekyll server and navigate to http://localhost:4000/.

However, when I run pandoc from `main_directory` to try and convert the markdown file to a PDF, I get the following error:

```
$ pandoc -t html rbenv/part-0/install-rbenv-content.md -o output.pdf --resource-path=assets/images

[WARNING] This document format requires a nonempty <title> element.
  Defaulting to 'install-rbenv-content' as the title.
  To specify a title, use 'title' in metadata or --metadata title="...".
Loading pages (1/6)
Warning: Failed to load file:///assets/images/2024-01-06-1037am.png (ignore)
Warning: Failed to load file:///assets/images/screenshot-2024-01-05-10-23-46.png (ignore)
Error: Failed to load file:///assets/images/2024-01-06-10.45am.png, with network status code 203 and http status code 0 - Error opening /assets/images/2024-01-06-10.45am.png: No such file or directory
Counting pages (2/6)                                               
Resolving links (4/6)                                                       
Loading headers and footers (5/6)                                           
Printing pages (6/6)
Done                                                                      
Exit with code 1 due to network error: ContentNotFoundError
Error producing PDF.
```

I also tried the `--extract-media` flag, and got a different error:

```
$ pandoc -t html rbenv/part-0/install-rbenv-content.md -o output.pdf --resource-path=../../ --extract-media=assets/images

pandoc: /assets/images/2024-01-06-1037am.png: withBinaryFile: does not exist (No such file or directory)
```

Judging by the `Failed to load file:///assets/images/` error, it looks like pandoc is not using the specified resource path, relative to the current directory (i.e. `main_directory`).  Not sure why. 

Can anyone help me out?

EDIT:

I tried removing the leading `/` from in front of `assets` inside the Markdown image declaration:

```
![some alt text](assets/images/screenshot-2024-01-05-10-23-46.png)
```

i.e. `(assets/images` instead of `(/assets/images`.  This time, I get a successful result:

```
$ pandoc -t html rbenv/part-0/install-rbenv-content.md -o output.pdf --resource-path=assets/images 
[WARNING] This document format requires a nonempty <title> element.
  Defaulting to 'install-rbenv-content' as the title.
  To specify a title, use 'title' in metadata or --metadata title="...".
Loading pages (1/6)
Counting pages (2/6)                                               
Resolving links (4/6)                                                       
Loading headers and footers (5/6)                                           
Printing pages (6/6)
Done                                      
```

However, now the images won't render in the browser.  I need a solution that lets me convert Markdown to PDF without having to manually remove that leading `/` from every image.

**Comment by jgm** [comment]
`/assets/images/screenshot-2024-01-05-10-23-46.png` is an absolute path, so pandoc is not going to look in resource path.  It does that only for relative paths.

In your case, if you can't use relative paths, you could try using a Lua filter that rewrites the paths as relative when you use pandoc.

**Comment by impostorsguides** [comment]
@jgm thanks for this.  I'm following the docs [here](https://pandoc.org/filters.html#but-i-dont-want-to-learn-haskell) and making a Python filter, which for now simply prints to STDOUT if the AST node is an `Image`:

```
#!/usr/bin/env python

"""
Pandoc filter to remove leading / from all markdown images 
"""

from pandocfilters import toJSONFilter, Image

def remove_leading_slash(key, value, format, meta):
  if key == 'Image':
    print(value)

if __name__ == "__main__":
  toJSONFilter(remove_leading_slash)
```

It's a "Hello, world" filter for now, just to make sure I'm doing things correctly up to this point.

I'm running pandoc using the following command:

```
pandoc -f markdown -t pdf install-rbenv-content.md -o output.pdf --filter ./image_converter_for_pdf.py
```

However, I'm getting the following error:

```
Error running filter ./image_converter_for_pdf.py:
Could not find executable python
```

I've got `python3` installed:

```
$ which python3
/opt/homebrew/bin/python3
```

And just in case pandoc is expecting the original `python` command, I've aliased `python3` to `python` in my shell.  I initially tried installing Python v2.7.0 to get a real `python` executable, but I got compilation errors.  I'm on a 2023 Apple M3 Max, and I suspect my hardware may be too new to support Python v2.

Any advice on the `Could not find executable python` error?

**Comment by jgm** [comment]
Try changing
```
#!/usr/bin/env python
```
to
```
#!/usr/bin/env python3
```
or if that doesn't work hard-code the path to python3.

My guess is that the environment in which the filters are being run won't include your local shell alias.

**Comment by impostorsguides** [comment]
I tried changing the shebang as you mentioned, but that didn't work.  I ended up using Paru to make both a converter and a filter, and that seems to have done the trick.

Thanks all the same for the quick response!
